File: packages/faro-ios-sdk/Sources/FaroSDK/Session/SessionStore.swift
import Foundation

internal struct StoredSession {
    let sessionId: String
    let startTime: Date
    let isSampled: Bool
}

internal final class SessionStore {
    private let defaults: UserDefaults

    private enum Keys {
        static let sessionId = "com.grafana.faro.session.id"
        static let startTime = "com.grafana.faro.session.startTime"
        static let isSampled = "com.grafana.faro.session.isSampled"
    }

    init(suiteName: String? = nil) {
        if let suiteName = suiteName {
            self.defaults = UserDefaults(suiteName: suiteName) ?? .standard
        } else {
            self.defaults = .standard
        }
    }

    func saveSession(_ session: StoredSession) {
        defaults.set(session.sessionId, forKey: Keys.sessionId)
        defaults.set(session.startTime.timeIntervalSince1970, forKey: Keys.startTime)
        defaults.set(session.isSampled, forKey: Keys.isSampled)
    }

    func loadSession() -> StoredSession? {
        guard let sessionId = defaults.string(forKey: Keys.sessionId) else { return nil }
        let startTime = Date(timeIntervalSince1970: defaults.double(forKey: Keys.startTime))
        let isSampled = defaults.bool(forKey: Keys.isSampled)
        return StoredSession(sessionId: sessionId, startTime: startTime, isSampled: isSampled)
    }

    func clear() {
        defaults.removeObject(forKey: Keys.sessionId)
        defaults.removeObject(forKey: Keys.startTime)
        defaults.removeObject(forKey: Keys.isSampled)
    }
}

File: packages/faro-android-sdk/src/main/kotlin/com/edkimmel/faro/session/SessionStore.kt
package com.edkimmel.faro.session

import android.content.Context
import android.content.SharedPreferences

data class StoredSession(
    val sessionId: String,
    val startTime: Long,
    val isSampled: Boolean
)

internal class SessionStore(context: Context) {
    private val prefs: SharedPreferences =
        context.getSharedPreferences("faro_session", Context.MODE_PRIVATE)

    companion object {
        private const val KEY_SESSION_ID = "session_id"
        private const val KEY_START_TIME = "start_time"
        private const val KEY_IS_SAMPLED = "is_sampled"
    }

    fun saveSession(session: StoredSession) {
        prefs.edit()
            .putString(KEY_SESSION_ID, session.sessionId)
            .putLong(KEY_START_TIME, session.startTime)
            .putBoolean(KEY_IS_SAMPLED, session.isSampled)
            .apply()
    }

    fun loadSession(): StoredSession? {
        val sessionId = prefs.getString(KEY_SESSION_ID, null) ?: return null
        val startTime = prefs.getLong(KEY_START_TIME, 0)
        val isSampled = prefs.getBoolean(KEY_IS_SAMPLED, true)
        return StoredSession(sessionId, startTime, isSampled)
    }

    fun clear() {
        prefs.edit().clear().apply()
    }
}
